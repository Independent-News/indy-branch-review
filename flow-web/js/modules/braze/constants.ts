export const BRAZE_EVENT_AUTORENEWAL_ON = 'turned_auto_renewalON';
export const BRAZE_EVENT_AUTORENEWAL_OFF = 'turned_auto_renewalOFF';
export const BRAZE_EVENT_PREMIUM_METER_EXPIRED = 'premium_meter_expired';
export const BRAZE_EVENT_REGISTRATION_COMPLETE = 'registration_complete';
export const BRAZE_EVENT_SUBSCRIPTION_START = 'subscription_start';
export const BRAZE_EVENT_SUBSCRIPTION_FINISH = 'subscription_finish';
export const BRAZE_EVENT_SUBSCRIPTION_ONBOARDING_SURVEY =
  'subscription_onboarding_survey';
export const BRAZE_EVENT_DONATION_START = 'donation_start';
export const BRAZE_EVENT_DONATION_FINISH = 'donation_finish';
