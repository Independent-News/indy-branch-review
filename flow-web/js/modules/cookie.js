/** @deprecated
 *
 * use getCookie() instead from app/public/js/utils/cookies.js
 */
export const getCookie = (name, defaultValue = null) => {
  const value = `; ${document.cookie}`;
  const parts = value.split(`; ${name}=`);
  return parts.length === 2
    ? decodeURIComponent(parts.pop().split(';').shift())
    : defaultValue;
};

/**
 * @param {string} name
 * @param {string} value
 * @param {object} [expire=false]
 * @param {number} [expire.days=0]
 * @param {number} [expire.hours=1]
 */
export const setCookie = (name, value, expire) => {
  const date = new Date();

  if (expire) {
    const expireTime = expire.days
      ? expire.days * 24 * 60 * 60 * 1000
      : expire.hours * 60 * 60 * 1000;
    date.setTime(date.getTime() + expireTime);
  }

  const parts = [];
  parts.push(`${name}=${encodeURIComponent(value)}`);
  parts.push('path=/');
  parts.push('secure');
  expire && parts.push(`expires=${date.toGMTString()}`);
  document.cookie = parts.join(';');
};

export const removeCookie = (name) => {
  const assignment = `${name}=null`;
  const expires = `expires=${new Date().toGMTString()}`;
  const path = 'path=/';
  document.cookie = [assignment, expires, path].join(';');
};

export const hasCookie = function (name) {
  return getCookie(name) !== null;
};
