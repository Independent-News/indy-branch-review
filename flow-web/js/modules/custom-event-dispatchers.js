import { dispatchCustomEvent } from '../utils/dispatchCustomEvent';

import { storeDigitalData } from './digital-data/digital-data-session';

const cacheSubscriptionInfo = ({
  subscription_length,
  subscription_price,
  subscription_package,
  selling_page_variant,
  subscribe_button_version,
}) => {
  storeDigitalData({
    subscription_length,
    subscription_price,
    subscription_package,
    selling_page_variant,
    subscribe_button_version,
  });
};

/**
 * @file
 * All custom events here are used for GTM.
 */

export const dispatchArticleBookmark = (data = {}) =>
  dispatchCustomEvent('article_bookmark', data);

export const dispatchAndCacheChangeSubscriptionPackage = (data = {}) => {
  cacheSubscriptionInfo(data);
  return dispatchCustomEvent('change_subscription_package', data);
};

export const dispatchPremiumArticleCTAPremiumClick = (data = {}) =>
  dispatchCustomEvent('premium_article_cta_premium', data);

export const dispatchPremiumArticleCTATideClick = (data = {}) =>
  dispatchCustomEvent('premium_article_cta_tide', data);

export const dispatchExpandablePromptCTAPremiumClick = (data = {}) =>
  dispatchCustomEvent('expandable_prompt_cta_premium', data);

export const dispatchExpandablePromptCTATideClick = (data = {}) =>
  dispatchCustomEvent('expandable_prompt_cta_tide', data);

export const dispatchChangePaymentMethod = (data = {}) =>
  dispatchCustomEvent('change_payment_method', data);

export const dispatchAndCacheSubscriptionButtonClick = (data = {}) => {
  cacheSubscriptionInfo(data);
  return dispatchCustomEvent('subscription_button_click', data);
};

export const dispatchFocusOnSubsRegField = (data = {}) =>
  dispatchCustomEvent('focus_reg_subscription_field', data);

export const dispatchDonation = (data = {}) =>
  dispatchCustomEvent('donation', data);

export const dispatchDonationApplePay = (data = {}) =>
  dispatchCustomEvent('donation_apple_pay', data);

export const dispatchDonationArticleBottom = (data = {}) =>
  dispatchCustomEvent('donation_article_bottom', data);

export const dispatchGalleryView = (data = {}) =>
  dispatchCustomEvent('gallery_view', data);

export const dispatchAutoGalleryView = (data = {}) =>
  dispatchCustomEvent('auto_gallery_view', data);

export const dispatchMarketingOpt = (data = {}) =>
  dispatchCustomEvent('marketing_opt', data);

export const dispatchRegistrationSuccess = (data = {}) =>
  dispatchCustomEvent('registration_success', data);

export const dispatchRegistrationSuccessStandard = (data = {}) =>
  dispatchCustomEvent('registration_success_standard', data);

export const dispatchRegistrationFailed = (data = {}) =>
  dispatchCustomEvent('registration_failed', data);

export const dispatchLoginSuccess = (data = {}) =>
  dispatchCustomEvent('login_success', data);

export const dispatchArticleShare = (data = {}) =>
  dispatchCustomEvent('article_share', data);

export const dispatchSearch = (data = {}) =>
  dispatchCustomEvent('search', data);

export const dispatchMegaMenuLinkClick = (data = {}) =>
  dispatchCustomEvent('mega_menu_link_click', data);

export const dispatchMainNavLinkClick = (data = {}) =>
  dispatchCustomEvent('main_nav_link_click', data);

export const dispatchMostPopularLinkClick = (data = {}) =>
  dispatchCustomEvent('most_popular_link_click', data);

export const dispatchPopularVideosLinkClick = (data = {}) =>
  dispatchCustomEvent('popular_videos_link_click', data);

export const dispatchSponsoredFeaturesLinkClick = (data = {}) =>
  dispatchCustomEvent('sponsored_features_link_click', data);

export const dispatchReadMoreLinkClick = (data = {}) =>
  dispatchCustomEvent('read_more_link_click', data);

export const dispatchIndy100TrendingLinkClick = (data = {}) =>
  dispatchCustomEvent('indy100_trending_link_click', data);

export const dispatchTaboolaFeedLinkClick = (data = {}) =>
  dispatchCustomEvent('taboola_feed_link_click', data);

export const dispatchRelatedArticlesLinkClick = (data = {}) =>
  dispatchCustomEvent('related_articles_link_click', data);

export const dispatchPaymentFormLoaded = (data = {}) =>
  dispatchCustomEvent('payment_form_loaded', data);

export const dispatchPaymentDetailsSuccess = (data = {}) =>
  dispatchCustomEvent('payment_details_success', data);

export const dispatchPaymentDetailsFailed = (data = {}) =>
  dispatchCustomEvent('payment_details_failed', data);

export const dispatchCommentSubmitted = (data = {}) =>
  dispatchCustomEvent('comment', data);

export const dispatchShowCommentsClick = (data = {}) =>
  dispatchCustomEvent('show_comments_click', data);

export const dispatchFlagCommentClick = (data = {}) =>
  dispatchCustomEvent('flag_comment_click', data);

export const dispatchVoteCommentClick = (data = {}) =>
  dispatchCustomEvent('vote_comment_click', data);

export const dispatchLoadMoreComments = (data = {}) =>
  dispatchCustomEvent('load_more_comments', data);

export const dispatchIpTabClick = (data = {}) =>
  dispatchCustomEvent('ip_tab_click', data);

export const dispatchIpEBookClick = (data = {}) =>
  dispatchCustomEvent('ip_ebook_click', data);

export const dispatchIpEventClick = (data = {}) =>
  dispatchCustomEvent('ip_event_click', data);

export const dispatchPushNotificationShown = (data = {}) =>
  dispatchCustomEvent('push_notification_shown', data);

export const dispatchPushNotificationOptIn = (data = {}) =>
  dispatchCustomEvent('push_notification_optin', data);

export const dispatchPushNotificationDismiss = (data = {}) =>
  dispatchCustomEvent('push_notification_dismiss', data);

export const dispatchLiveBlogBulletedListClick = (data = {}) =>
  dispatchCustomEvent('liveblog_bulleted_list_click', data);

export const dispatchArticleBulletedListClick = (data = {}) =>
  dispatchCustomEvent('article_bulleted_list_click', data);

export const dispatchBlogPostArticleLinkClick = (data = {}) =>
  dispatchCustomEvent('blog_post_article_link_click', data);

export const dispatchRegistrationWallClick = (data = {}) =>
  dispatchCustomEvent('registration_wall_click', data);

export const dispatchMostWatchedLoad = (data = {}) =>
  dispatchCustomEvent('most_watched_load', data);

export const dispatchMostWatchedClick = (data = {}) =>
  dispatchCustomEvent('most_watched_click', data);

export const dispatchContextualInlineClick = (data = {}) =>
  dispatchCustomEvent('contextual_inline_click', data);

export const dispatchContextualInlineTopicClick = (data = {}) =>
  dispatchCustomEvent('contextual_inline_topic_click', data);

export const dispatchContextualInlineNonTopicClick = (data = {}) =>
  dispatchCustomEvent('contextual_inline_non_topic_click', data);

export const dispatchLiveBlogKeyPointsClick = (data = {}) =>
  dispatchCustomEvent('liveblog_key_points_click', data);

export const dispatchSellingPageLoaded = (data = {}) =>
  dispatchCustomEvent('selling_page_loaded', data);

export const dispatchMostCommentedSectionClick = (data = {}) =>
  dispatchCustomEvent('most_commented_section_click', data);

export const dispatchUpdateComponentsList = (data = {}) =>
  dispatchCustomEvent('update_components_list', data);

export const dispatchTopicListItemClick = (data = {}) =>
  dispatchCustomEvent('article_topic_tag_click', data);

export const dispatchAdBlockDetected = () =>
  dispatchCustomEvent('adblock_detected');

export const dispatchExperienceExecuted = (data = {}) =>
  dispatchCustomEvent('experience_executed', data);

export const dispatchVoucherTickerLinkClick = (data = {}) =>
  dispatchCustomEvent('voucher_ticker_link_click', data);

export const dispatchVoucherTickerArrowClick = (data = {}) =>
  dispatchCustomEvent('voucher_ticker_arrow_click', data);

export const dispatchBurgerMenuClick = () =>
  dispatchCustomEvent('burger_menu_click');

export const dispatchEditionLinkClick = (data = {}) =>
  dispatchCustomEvent('edition_link_click', data);

export const dispatchTempLinkClick = (data = {}) =>
  dispatchCustomEvent('temporary_link_click', data);

export const dispatchAffiliateLinkClick = (data = {}) =>
  dispatchCustomEvent('affiliate_link_click', data);

export const dispatchWebpushSignupClick = () =>
  dispatchCustomEvent('webpush_signup_click');

export const dispatchNotificationPromptAccept = () =>
  dispatchCustomEvent('notification_prompt_accept');

export const dispatchNotificationPromptDismiss = () =>
  dispatchCustomEvent('notification_prompt_dismiss');

export const dispatchNotificationPromptLoad = (data = {}) =>
  dispatchCustomEvent('notification_prompt_load', data);

// Price comparison

export const dispatchPriceComparisonListBuyNowClick = (data = {}) =>
  dispatchCustomEvent('price_comparison_list_buy_now_click', data);

export const dispatchPriceComparisonButtonsBuyNowClick = (data = {}) =>
  dispatchCustomEvent('price_comparison_buttons_buy_now_click', data);

// Product carousel

export const dispatchProductCarouselLoaded = (data = {}) =>
  dispatchCustomEvent('product_comparison_carousel_load', data);

export const dispatchProductCarouselBuyNowClick = (data = {}) =>
  dispatchCustomEvent('product_carousel_buy_now_click', data);

export const dispatchProductCarouselReadReviewClick = (data = {}) =>
  dispatchCustomEvent('product_carousel_read_review_click', data);

// Product comparison

export const dispatchProductComparisonCheckboxClick = (data = {}) =>
  dispatchCustomEvent('product_comparison_checkbox_click', data);

export const dispatchProductComparisonModalOpen = (data = {}) =>
  dispatchCustomEvent('product_comparison_modal_open', data);

export const dispatchProductComparisonBuyNowClick = (data = {}) =>
  dispatchCustomEvent('product_comparison_modal_buy_now_click', data);

export const dispatchProductComparisonReadReviewClick = (data = {}) =>
  dispatchCustomEvent('product_comparison_modal_read_review_click', data);

// Cancellation

export const dispatchAccountOverviewTabMyDetailsClick = (data = {}) =>
  dispatchCustomEvent('account_overview_tab_my_details', data);

export const dispatchAccountOverviewTabChangePasswordClick = (data = {}) =>
  dispatchCustomEvent('account_overview_tab_change_password', data);

export const dispatchAccountOverviewTabPaymentAndBillingClick = (data = {}) =>
  dispatchCustomEvent('account_overview_tab_payments_and_billing', data);

export const dispatchAccountOverviewTabManageSubscriptionClick = (data = {}) =>
  dispatchCustomEvent('account_overview_tab_manage_subscription', data);

export const dispatchAccountCancellationBarrier1Click = (data = {}) =>
  dispatchCustomEvent('account_cancellation_barrier_1', data);

export const dispatchAccountCancellationBarrier2Click = (data = {}) =>
  dispatchCustomEvent('account_cancellation_barrier_2', data);

export const dispatchAccountCancellationKeepSubscriptionClick = (data = {}) =>
  dispatchCustomEvent('account_cancellation_keep_subscription', data);

export const dispatchAccountCancellationTakeThisOfferClick = (data = {}) =>
  dispatchCustomEvent('account_cancellation_take_this_offer', data);

export const dispatchAccountCancellationContinueToPaymentClick = (data = {}) =>
  dispatchCustomEvent('account_cancellation_continue_to_payment', data);

export const dispatchAccountCancellationReactivateSubscriptionClick = (
  data = {},
) => dispatchCustomEvent('account_cancellation_reactivate_subscription', data);

export const dispatchStageBarrier1Click = (data = {}) =>
  dispatchCustomEvent('stage_barrier_1', data);

export const dispatchStageBarrier2Click = (data = {}) =>
  dispatchCustomEvent('stage_barrier_2', data);

export const dispatchSupportNSCLinkClick = (data = {}) =>
  dispatchCustomEvent('support_nsc_link_click', data);

export const dispatchSupportNowEncoreClick = (data = {}) =>
  dispatchCustomEvent('support_now_encore_click', data);
