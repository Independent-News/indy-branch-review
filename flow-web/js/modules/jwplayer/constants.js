export const PARAM_AD_POS = 'ap';
