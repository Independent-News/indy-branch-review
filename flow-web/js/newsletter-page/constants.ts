export const EMAIL_INPUT_FIELD_ID = 'email-input-field';
export const ERROR_MESSAGE_ID = 'error-message';
export const INPUT_WIDGET_ID = 'input-widget';
export const OFFERS_CHECKBOX_ID = 'receiveIndyOffers';
export const SUBMIT_BUTTON_ID = 'submit-newsletter-form';
export const SUBMIT_BUTTON_TEXT_ID = 'submit-newsletter-button-text';
export const SUCCESS_POPUP_ID = 'success-popup';
