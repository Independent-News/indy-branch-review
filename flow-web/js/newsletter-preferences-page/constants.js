export const NEWSLETTER_TITLE = 'newsletter-title';
