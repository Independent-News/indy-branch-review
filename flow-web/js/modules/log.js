/* global Sentry */
/* eslint-disable no-console */

import { PROD } from '#app/constants/server';

import { getCookie } from '#app/public/js/utils/cookie';

import { isSentryEnabledClientSide } from '../utils/sentry';

let showLog;
let showSentry;

const shouldLog = () => {
  if (typeof window === 'undefined') {
    return process.env.SERVER !== PROD;
  }

  if (typeof window !== 'undefined' && typeof showLog === 'undefined') {
    const params = new URLSearchParams(window?.location?.search);
    showLog = params.has('__DEBUG__');
  }

  return showLog || getCookie('__DEBUG__') === 'true';
};

const shouldLogToSentry = () => {
  if (typeof showSentry === 'undefined') {
    showSentry = isSentryEnabledClientSide();
  }

  return showSentry;
};

export const group = (...args) => {
  const showLog = shouldLog();

  showLog && console.group(...args);
};

export const groupEnd = (...args) => {
  const showLog = shouldLog();

  showLog && console.groupEnd(...args);
};

export const log = (...args) => {
  const showLog = shouldLog();

  showLog && console.log(...args);
};

export const warn = (...args) => {
  const showLog = shouldLog();
  const showSentry = shouldLogToSentry();

  showLog && console.warn(...args);
  showSentry && Sentry.captureException(...args);
};

export const error = (...args) => {
  const showLog = shouldLog();
  const showSentry = shouldLogToSentry();

  showLog && console.error(...args);
  showSentry && Sentry.captureException(...args);
};
