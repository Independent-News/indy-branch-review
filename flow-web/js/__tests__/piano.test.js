/**
 * @jest-environment jsdom
 */

import { COOKIE_SUBSCRIBER } from '#app/constants/cookies';

import { TRACKING_PAYMENT_KEYS } from '#app/constants/adobeConstants';
import { CONSENT_TO_PAY_ORIGINAL_PAGE_REDIRECT_URL } from '#app/constants/localStorage';
import {
  SUBSCRIPTION_TRACKING_NAME_ONE_OFF_DONATION,
  SUBSCRIPTION_TRACKING_NAME_RECURRING_DONATION,
} from '#app/constants/subscription';

import {
  handleAlreadyHasAccessForSubscriptions,
  handleCheckoutCompleteForSubscriptions,
} from '#app/public/js/piano';

import { setCookie } from '../modules/cookie';
import { dispatchPaymentDetailsSuccess } from '../modules/custom-event-dispatchers';
import {
  getDigitalDataByKeys,
  removeDigitalDataByKeys,
} from '../modules/digital-data/digital-data-session';
import { error } from '../modules/log';
import * as thatModule from '../piano';
import { redirect } from '../utils/redirect';

jest.mock('../utils/redirect');
jest.mock('../modules/cookie');
jest.mock('../modules/digital-data/digital-data-session');
jest.mock('../modules/custom-event-dispatchers');
jest.mock('../modules/taboola');
jest.mock('../utils/internalApi');
jest.mock('../modules/log');

describe('Piano public JS methods', () => {
  beforeEach(() => {
    window.tp = [];
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.useRealTimers();
  });

  afterAll(() => {
    delete window.tp;
    delete global.fetch;
    jest.restoreAllMocks();
  });

  describe('setSubscriberCookieAndRedirectToConfirmationPage()', () => {
    it('should set subscriber cookie with correct expiry date and redirect to confirmation page', async () => {
      thatModule.setSubscriberCookieAndRedirectToConfirmationPage();
      expect(setCookie).toHaveBeenCalledTimes(1);
      expect(setCookie).toHaveBeenCalledWith(COOKIE_SUBSCRIBER, true, {
        days: thatModule.EXPIRE_DAYS,
      });
      jest.advanceTimersByTime(100);
      expect(redirect).toHaveBeenCalledTimes(1);
      expect(redirect).toHaveBeenCalledWith('/subscribe/confirmation');
    });
  });

  describe('handleCheckoutCompleteForSubscriptions()', () => {
    beforeAll(() => {
      jest
        .spyOn(thatModule, 'setSubscriberCookieAndRedirectToConfirmationPage')
        .mockImplementation();
    });

    afterAll(() => {
      thatModule.setSubscriberCookieAndRedirectToConfirmationPage.mockRestore();
    });

    it('should expose correct checkout complete event handler to Piano SDK which sets subscriber cookie and redirects to confirmation page', () => {
      handleCheckoutCompleteForSubscriptions();
      expect(window.tp).toMatchInlineSnapshot(`
        [
          [
            "addHandler",
            "checkoutComplete",
            [MockFunction],
          ],
        ]
      `);
      const checkoutCompleteHandle = window.tp[0][2];
      checkoutCompleteHandle();
      expect(
        thatModule.setSubscriberCookieAndRedirectToConfirmationPage,
      ).toHaveBeenCalledTimes(1);
    });
  });

  describe('handleAlreadyHasAccessForSubscriptions()', () => {
    beforeAll(() => {
      Object.defineProperty(window, 'localStorage', {
        value: {
          getItem: jest.fn(),
          removeItem: jest.fn(),
        },
        writable: true,
      });
    });

    it('will do nothing if a different event is emitted to that of already has access', () => {
      handleAlreadyHasAccessForSubscriptions();
      const customEventHandler = window.tp[0][2];
      customEventHandler({ eventName: 'some-other-event' });
      expect(redirect).not.toHaveBeenCalled();
      expect(removeDigitalDataByKeys).not.toHaveBeenCalled();
      expect(window.localStorage.getItem).not.toHaveBeenCalled();
      expect(window.localStorage.removeItem).not.toHaveBeenCalled();
    });

    it('will expose correct custom event handler to Piano SDK which removes relevant digital data and redirects to subscribe page on has access close event if no consent or pay original url is cached', () => {
      window.localStorage.getItem.mockReturnValueOnce(null);
      handleAlreadyHasAccessForSubscriptions();
      expect(window.tp).toMatchInlineSnapshot(`
        [
          [
            "addHandler",
            "customEvent",
            [Function],
          ],
        ]
      `);
      const customEventHandler = window.tp[0][2];
      customEventHandler({ eventName: 'close-has-access' });
      expect(redirect).toHaveBeenCalledTimes(1);
      expect(redirect).toHaveBeenCalledWith('/subscribe');
      expect(removeDigitalDataByKeys).toHaveBeenCalledTimes(1);
      expect(removeDigitalDataByKeys).toHaveBeenCalledWith(
        TRACKING_PAYMENT_KEYS,
      );
      expect(window.localStorage.getItem).toHaveBeenCalledTimes(1);
      expect(window.localStorage.getItem).toHaveBeenCalledWith(
        CONSENT_TO_PAY_ORIGINAL_PAGE_REDIRECT_URL,
      );
      expect(window.localStorage.removeItem).not.toHaveBeenCalled();
    });

    it('will remove relevant digital data and redirect to consent or pay original url if cached in local storage', () => {
      const mockCachedUrl = '/some-url';
      window.localStorage.getItem.mockReturnValueOnce(mockCachedUrl);
      handleAlreadyHasAccessForSubscriptions();
      expect(window.tp).toMatchInlineSnapshot(`
        [
          [
            "addHandler",
            "customEvent",
            [Function],
          ],
        ]
      `);
      const customEventHandler = window.tp[0][2];
      customEventHandler({ eventName: 'close-has-access' });
      expect(redirect).toHaveBeenCalledTimes(1);
      expect(redirect).toHaveBeenCalledWith(mockCachedUrl);
      expect(removeDigitalDataByKeys).toHaveBeenCalledTimes(1);
      expect(removeDigitalDataByKeys).toHaveBeenCalledWith(
        TRACKING_PAYMENT_KEYS,
      );
      expect(window.localStorage.removeItem).toHaveBeenCalledTimes(1);
      expect(window.localStorage.removeItem).toHaveBeenCalledWith(
        CONSENT_TO_PAY_ORIGINAL_PAGE_REDIRECT_URL,
      );
    });

    it.each([
      SUBSCRIPTION_TRACKING_NAME_ONE_OFF_DONATION,
      SUBSCRIPTION_TRACKING_NAME_RECURRING_DONATION,
    ])(
      'will remove relevant digital data and redirect to donations page if tracking name is %s',
      (trackingName) => {
        jest
          .spyOn(thatModule, 'getResourceTrackingName')
          .mockReturnValueOnce(trackingName);

        window.localStorage.getItem.mockReturnValueOnce(null);

        handleAlreadyHasAccessForSubscriptions();

        expect(window.tp).toMatchInlineSnapshot(`
          [
            [
              "addHandler",
              "customEvent",
              [Function],
            ],
          ]
        `);
        const customEventHandler = window.tp[0][2];
        customEventHandler({ eventName: 'close-has-access' });

        expect(redirect).toHaveBeenNthCalledWith(1, '/donate');
      },
    );
  });

  describe('handleCheckoutComplete()', () => {
    it('will expose correct checkout complete callback to Piano SDK', () => {
      thatModule.handleCheckoutComplete();
      expect(window.tp).toMatchInlineSnapshot(`
        [
          [
            "addHandler",
            "checkoutComplete",
            [Function],
          ],
        ]
      `);
      const [[, , callback]] = window.tp;
      callback();
      jest.advanceTimersByTime(100);
      expect(redirect).toHaveBeenCalledTimes(1);
      expect(redirect).toHaveBeenCalledWith(
        '/subscribe/confirmation?save_offer=true',
      );
    });
  });

  describe('showPianoOffer()', () => {
    beforeEach(() => {
      window.tp.offer = {
        show: jest.fn(),
      };
      jest.spyOn(thatModule, 'addHandlers').mockImplementation();
    });

    it('will add handlers to Piano SDK and expose correct show offer callback to be called once initialised', () => {
      thatModule.showPianoOffer({
        offerId: 'offerId',
        termId: 'termId',
        trackingData: 'trackingData',
        promoCode: 'promoCode',
        showCloseButton: false,
      });
      expect(window.tp).toMatchSnapshot();
      const [[, callback]] = window.tp;
      callback();
      expect(window.tp.offer.show).toHaveBeenCalledTimes(1);
      expect(window.tp.offer.show).toHaveBeenCalledWith({
        offerId: 'offerId',
        termId: 'termId',
        showCloseButton: false,
        promoCode: 'promoCode',
      });
    });
  });

  describe('handleCheckoutCompleteForDonations()', () => {
    const originalFetch = global.fetch;

    beforeAll(() => {
      global.fetch = jest.fn();
    });

    afterAll(() => {
      global.fetch = originalFetch;
    });

    it('will expose correct checkout complete event handler to Piano SDK which tracks successful payment, updates logged in cookies and redirects to donations page', async () => {
      const mockTrackingData = 'some-tracking-data';
      getDigitalDataByKeys.mockReturnValueOnce(mockTrackingData);
      global.fetch.mockResolvedValueOnce({
        ok: true,
        json: jest.fn().mockResolvedValueOnce({}),
      });
      thatModule.handleCheckoutCompleteForDonations();
      expect(window.tp).toMatchInlineSnapshot(`
        [
          [
            "addHandler",
            "checkoutComplete",
            [Function],
          ],
        ]
      `);
      const checkoutCompleteHandler = window.tp[0][2];
      await checkoutCompleteHandler();
      expect(getDigitalDataByKeys).toHaveBeenCalledTimes(1);
      expect(getDigitalDataByKeys).toHaveBeenCalledWith(TRACKING_PAYMENT_KEYS);
      expect(dispatchPaymentDetailsSuccess).toHaveBeenCalledTimes(1);
      expect(dispatchPaymentDetailsSuccess).toHaveBeenCalledWith(
        mockTrackingData,
      );
      expect(error).not.toHaveBeenCalled();
      expect(redirect).toHaveBeenCalledTimes(1);
      expect(redirect).toHaveBeenCalledWith('/donate/thank-you');
    });

    it("will only console error but still execute remaining functionality if fails to update user's logged in cookies", async () => {
      const mockErrorMessage = 'some-error-message';
      global.fetch.mockResolvedValueOnce({
        ok: false,
        json: jest.fn().mockResolvedValueOnce({
          message: mockErrorMessage,
        }),
      });
      thatModule.handleCheckoutCompleteForDonations();
      const checkoutCompleteHandler = window.tp[0][2];
      await checkoutCompleteHandler();
      expect(error).toHaveBeenCalledTimes(1);
      expect(error).toHaveBeenCalledWith(
        "Error: failed to update donator's logged in cookies,",
        new Error(mockErrorMessage),
      );
      expect(getDigitalDataByKeys).toHaveBeenCalledTimes(1);
      expect(dispatchPaymentDetailsSuccess).toHaveBeenCalledTimes(1);
      expect(redirect).toHaveBeenCalledTimes(1);
      expect(redirect).toHaveBeenCalledWith('/donate/thank-you');
    });
  });
});
