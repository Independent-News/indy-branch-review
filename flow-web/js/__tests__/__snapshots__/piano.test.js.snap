// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Piano public JS methods showPianoOffer() will add handlers to Piano SDK and expose correct show offer callback to be called once initialised 1`] = `
[
  [
    "init",
    [Function],
  ],
]
`;
